### ISSUE TITLE
Widget support

### ISSUE THREAD

**Issue by Shinra04** [8bitPit/Niagara-Issues#117]
Allow widgets with minimalist icons on the top-right portion of screen.

**Comment by zbhoy** [comment]
I was thinking a page to the right with a list of active widgets would be cool.

**Comment by Shinra04** [comment]
Oh yes, but I was thinking how to keep it minimalistic.

**Comment by jakcharvat** [comment]
I wouldn't close this issue so we can keep track of the idea 😉

**Comment by Shinra04** [comment]
Eh, okay. But my #131 thread has a different suggestion for where the widgets should be. The suggestion I put here would make the launcher look a bit cluttered

**Comment by oliver-cfc** [comment]
Just wanted to say that lack of widget support is the only thing stopping me from uninstalling SmartLauncher and moving to this full time.  I really miss being able to replace the time/date on the main screen with a weather widget that tells me the temps for the next few hours etc.
I think having the time there is redundant since it's also in the top bar anyway

**Comment by preetangad** [comment]
Really loving this launcher, first one I've made default ever. I just want to be able to use a couple of widgets, especially a calendar widget 2x4 and some shortcuts like Google maps and reminders (all 1x1).

**Comment by 8bitPit** [comment]
@preetangad Shortcut support is an independent issue, they're easier to implement, see #62

**Comment by 8bitPit** [comment]
but I know that other launchers display them in the widgets section

**Comment by vbujanca** [comment]
I'd like a dedicated widget page (or more) but also top favourites page

**Comment by hasarmu** [comment]
Hide the Clock Option! Then You can Set a Widget With Full Width! Some People Really Don't need a clock bruh!

**Comment by prattsp** [comment]
It should allow to lock the screen, if possible just by touching the blank space on home screen.

Also more importantly it should allow us to add the emergency details to add on lock screen.

**Comment by 8bitPit** [comment]
Praveen Chandran wrote in Telegram:

> I have an idea for integrating widgets. The top portion that currently has the date and time will be a horizontally scrollable page like thing. So that we can add widget into each page. It's the top area so that it won't interfere with current the swipe gestures

**Comment by jujijoog** [comment]
Great launcher. Impressed. Definitely just waiting on at least some widget support before I commit to it. Keep up the good work!

**Comment by preetangad** [comment]
I love the idea of being able to put one widget on the top right, especially if it's more than 1x1. I really wish I can see my calendar without clicking anything and I'm sure other people have their own things too.

**Comment by ShinobiWPS** [comment]
> I was thinking a page to the right with a list of active widgets would be cool.

+1
To avoid the calendar widget it could be enough (waiting opinion from everybody) to have a persistent expanded Height-space if you make Favourite your calendar app.

using current behaviour, You see events notifications in the launcher from your calendar for the current day only BUT if you press on it, the list of events won't disappear from the launcher Icon.

possible drawback if you have plenty of events for the current day only the list will move favourite apps below it even more.

possible solution: limit the list to 3 rows height and make it scrollable(or not)

**Comment by ShinobiWPS** [comment]
ANOTHER suggestion would be to show ONLY reminders  of your calendar persistently  for the current day only with the above technique.
They can be visually replaced if you have new notification /events in your calendar and once dismissed you will see your list of reminders for the current day only.

**Comment by ShinobiWPS** [comment]
We are trying to emulate Google usability engineers.

**Comment by CiriousJoker** [comment]
I'd suggest putting the Widgets in the usual page based system when the user swipes from left to right.

**Comment by Guyzers** [comment]
Why not just put a Widget Icon in the up/down scroll area? That way, Widgets don't have to interfere with the apps and the screen still would be less clustered

**Comment by Shinra04** [comment]
Having the widgets in the upper section of the screen makes it a one-tap access to a couple, and one gesture away from the rest. I doubt people will have more than about 7 widgets they'd really need.

**Comment by tjsuominen** [comment]
Let Niagara be unique (in a good way) and not too "copy-cat" (just saying, nothin' more). Yet repeating myself, missing widgets support from home-screen is the single biggest thing for me preventing Niagara to become my daily driver (and "At a Glance" like calendar feature).

**Comment by provalinf** [comment]
Hi, guys, I just discovered this launcher, it's really great.
But like many, it is missing the possibility to add a page to put widgets only, for me the most important thing are the widgets for the agendas, rss feeds and galleries.

Thanks to the dev

**Comment by Dran01** [comment]
Is there any way to get an additional home page dedicated for widgets using swipe gestures?? 

I really loved your launcher and please don't stop developing itt!!!!

**Comment by chriscam85** [comment]
Guys, just add another launcher app to your favorites. I use Evie, and put all my widgets there. Then home button is back to Niagara. Perfect. No need for extra development.

**Comment by Guyzers** [comment]
The best way (in my perspective) is to use the iOS like way to get to the widgets using the left side of the screen. Since some of us are lefties, maybe have a toggle which switches the gestures.

**Comment by tjsuominen** [comment]
Absodefinitely NOT to have widgets *only* on a separate screen, but on the default (home) screen. If one thing I think Niagara should awp ("adopt with pride") from Nova Launcher, is the way it handles widgets - place freely on the (home) screen. Yet, understanding how Niagara works, widget(s) might need to have a dedicated (fixed?) position on the screen - or perhaps not. It all depends HOW Niagara bahavior is developed. For me, it's truly a showstopper NOT having "at glance" like info (calendar / next meeting) in sight all the time.

**Comment by oliver-cfc** [comment]
> 
> 
> Absodefinitely NOT to have widgets _only_ on a separate screen, but on the default (home) screen. If one thing I think Niagara should awp ("adopt with pride") from Nova Launcher, is the way it handles widgets - place freely on the (home) screen. Yet, understanding how Niagara works, widget(s) might need to have a dedicated (fixed?) position on the screen - or perhaps not. It all depends HOW Niagara bahavior is developed. For me, it's truly a showstopper NOT having "at glance" like info (calendar / next meeting) in sight all the time.

I felt the same way but there are apps in the playstore that will overlay the widget(s) of your choice but only while specific apps are running in the foreground.  This is how I'm placing a weather widget just under the notification bar and only when I'm on the homescreen.  If I start any other app then the widget disappears. 
 I'd rather have native support from Niagara but it's working in the meantime.

The one I'm using is called overlays.  I'm not affiliated in any way with it and there have been a couple of bugs/annoyances but it's good enough for me.

**Comment by chareos** [comment]
What about offer (long tap) to replace an app entry with it's 4x1 / 4x2 widget, if available ?
It would stay consistent with the music widget appearance and usage model, no need for a separate home page or different scrolling logic to get to a separate widget space.

**Comment by verboze** [comment]
Similar to oliver-cfc, I use this launcher in conjunction with Popup Widget, Tasker, and Gesture Navigations; I have mapped a gesture to launch that Popup Widget. Prob solved for me.

**Comment by chriscam85** [comment]
I have a really nice combination Niagara and Kustom Live wallpaper KLWP going. You can basically create your own/download widgets in KLWP. The only issue I'm having is I can't use the Touch Widget capabilities because perhaps Niagara doesn't have "Tap to Wallpaper must be supported by the Launcher"??? Any way to check this?

**Comment by johnhayesio** [comment]
> Similar to oliver-cfc, I use this launcher in conjunction with Popup Widget, Tasker, and Gesture Navigations; I have mapped a gesture to launch that Popup Widget. Prob solved for me.

This worked perfectly for the time being until native support is added for widgets. Thanks for the Popup Widget suggestion.

**Comment by tjsuominen** [comment]
Still trying to convince ;) for the importance of the "Next agenda item" and the ability to have one placed for the Niagara main home screen with a "At glance" like functionality. Would satisfy many, many users. And I think I could fit just nicely from designwise as well... :)

**Comment by cauerego** [comment]
i used niagara for a while and fell in love like everyone else here...

a big reason is i couldn't care less about widgets.

but once pie hit my nokia 6.1 i had to go back to the default pixel launcher.

as many mentioned, i was of course always missing the calendar smart alerts and smart app suggestions... but also app shortcuts, such as chrome web apps and maps friends locations. mainly, though, i couldn't stand the lack of integration from the new task switcher, by swiping up from the bottom. i think it's broken in any other launcher. probably android's fault.

and i love the new assistance button that goes in place of the app switcher, to get soft lock native among other nice utilities there, so i'm not going back despite of being a bit clumsy to swap on previous app.

in any case, to me this was the solution:

i now keep niagara as an app on the home screen.

just to go through apps when needed, instead of using the freaking useless default app list.

this way we get the best of both worlds, without needing to install any extra clumsy memory consumption apps (only thing is a bit less of convenience there). memory is essential if you're in a cheap device. and it works! :)

**Comment by piekay** [comment]
Another way to handle this would be to swipe from Widget to Widget (left/right swipe)

**Comment by bala7s** [comment]
#741: Please add widget support, but with a complete separate screen that only usable for widgets.
This should be on the top of home screen, so there would be 3 screens from top to bottom as widget screen would be on the top, home screen with favorite apps in the middle, and the all apps on the bottom. And a single vertical swipe would allow the user to switch between them.
I think this would make a much cleaner interface than putting everything into one singe screen.

**Comment by 8bitPit** [comment]
I am not really a fan of widget pages. The only thing widgets have in common there is, well, they’re all widgets. So why should they be grouped together? Showing a weather widget below the clock and accessing your calendar agenda via a swipe / by tapping on a button makes more sense to me. Also, widgets often tend to use much ram. The concept of a widget list would kind of endorse user to cluster the page full of widgets. A swipe menu shows only a single widget at once, so not every widget has to be loaded into ram when you want to access one. 
That’s why I’m planning to add the following integration: 
You can place a single widget below the clock or replace it for showing weather and / or time, or an at a glance widget.
More widgets can be accessed by swiping right on an app.
Another idea would be that you can customize the google search button at the bottom right, so it can also launch a widget, e.g. your calendar.

**Comment by Shinra04** [comment]
![screenshot_20190216-011005](https://user-images.githubusercontent.com/43073817/52880470-12121900-3188-11e9-84ed-d2db6c90eb7b.png)
I've been using the app Overlays to meet my widget needs, but it's just a temporary replacement. The placement of the widgets(top-right and bottom) is perfectly accessible and doesn't intrude with other elements. Perhaps this gives a good idea on where to place them.

**Comment by tjsuominen** [comment]
@8bitPit ; 

Huge thumbs up for this:

> You can place a single widget below the clock or replace it for showing weather and / or time, or an at a glance widget.

With big support to At a Glance widget (which I understood you'll make a "Niagara one?).

**Comment by Dedisurya** [comment]
I need only one home screen widget. Only one home screen widget is enough for me. And i dont need have a download widget, which is show what we download, i dont need that. Just home screen and menu.
The concept which Niagara Launcher do now is perfect for me. Simple, ideal, useful, efective, and efficient.

**Comment by Dedisurya** [comment]
In appearance, I think this is enough. but there are still some problems in the Niagara launcher performance. I open the music application, it usually happens to pop up the music control menu automatically to the home screen. but when I used Niagara launcher, the pop up menu didn't appear. it would be better if it was corrected. Thank you

**Comment by Dedisurya** [comment]
What for the widget is ?.
Home screen and menu was enough

**Comment by paijojo** [comment]
only need the home screen and menu. there is no need to have many screens for favorites or screens for newly downloaded applications or whatever. only need the home screen and menu. there is no need for a search page, the home button or google application is enough to do a search. do not make many rights stacked only for the same function. simplicity is everything.

**Comment by paijojo** [comment]
what the Niagara launcher has done so far is for the home screen display and the menu is correct.

**Comment by aatifsyed** [comment]
Swiping L→R opens notifications
Pressing and holding opens options
Why not have swiping L←R open a widget?
You could take cues from action launcher in terms of selecting a widget for an app

**Comment by 8bitPit** [comment]
> Why not have swiping L←R open a widget?

Because that gesture dismisses notifications. However, I think a widget could also fit in the notifications/shortcuts menu.

**Comment by TheSophiist** [comment]
Can we get align postion for widgets, i mean aligning them to left, right or centre?

**Comment by 8bitPit** [comment]
@TheSophiist 
For the homescreen widget there will be free different alignment modes:
![image](https://user-images.githubusercontent.com/9284845/57967525-7c418200-7960-11e9-82b6-6f61ecb183cf.png)

**Comment by fante76** [comment]
I'd like the full option

Il sab 18 mag 2019, 11:31 Peter <notifications@github.com> ha scritto:

**Comment by JebDslayer** [comment]
Perhaps a page you can swipe to that you can swipe through selected widgets like Google now cards?

**Comment by drc0369** [comment]
Probably the biggest for me would be to incorporate the Settings widget to add to the launcher. I use the WiFi and Bluetooth settings widget for one press access to these two important settings for when I need to turn them on and off. Swiping down from the top is easy, but it only turns them on and off, not select a certain network or device.

**Comment by 8bitPit** [comment]
@drc0369 you can hold on the WiFi or Bluetooth button to access further options. If you want to access these options directly from Niagara, I recommend searching for an app that let's you place a widget for that type of functionality. Unfortunately I haven't used such an app yet, so I can't recommend one.

**Comment by preetangad** [comment]
Thanks so much for adding widget support! This is exactly what I was
looking for.

Quick request: can you also allow the ability to use widgets from the Work
Profile? Want to use the Calendar widget to see my next meeting but my
company uses work profiles. I think pixel launcher supports them so it
should be possible.

**Comment by 8bitPit** [comment]
@preetangad glad you like it 😃 I'll try to support them.

**Comment by tjsuominen** [comment]
To repeat (?) myself, any chance having BOTH weather info (as widget) and next meeting ("At a glance") supported at the same time?

**Comment by 8bitPit** [comment]
@tjsuominen additional widgets can probably be placed in app swipe menus, but not on the homescreen. However, The "At a glance" widget might be integrated into the Niagara's time & date widget: #55

**Comment by tjsuominen** [comment]
@8bitPit Appreciated! I think the "Next meeting" aka "At a glance" point is being noted 'At a glance' = deserves (read: needs :o) to be on the homescreen. Strong +1 for the #55 and shall continue feedback @ there. Thx!

**Comment by 8bitPit** [comment]
@preetangad figured out that Niagara already supports work profile widgets. However, some shortcuts are also labeled as widgets in other launchers. Those will be supported in the future: https://github.com/8bitPit/Niagara-Issues/issues/62

**Comment by niakoi1** [comment]
Just some feedback on the new widget feature. When in edit/resize mode the widget let's you drag it up and down by holding down on the upper bar's arrows.I'm not sure why this is needed, unless more than one widget can be added and repositioning needed, which is currently not  the case. Anyhow, the way it's now, you can drag it all the way down and thus create an unnecessary gap at the top that pushes all home apps to the bottom. Or you can drag it all the way up, but this pushes  the clock off the top of the screen. Again not very desirable.Maybe I missed something, but right now dragging makes no sense either up or down and just creates issues.

**Comment by chriscam85** [comment]
@niakoi1 I use this feature to show my live background. I don't want that section covered, and this helps me position the widget correctly.

**Comment by niakoi1** [comment]
Not sure what you are saying.
I pointed out that dragging it up pushes the clock off the top of the
screen. (clock and widget are incorrectly positioned)
And dragging it down creates a gap (empty screen space) on the screen.
(clock and widget are incorrectly positioned)
[image: Screenshot_20190816-154054.png]
[image: Screenshot_20190816-154116.png]
[image: Screenshot_20190816-154039.png]

**Comment by Rod911** [comment]
@niakoi1 the clock going off the screen is not intended and will be fixed. The upper slider is used to adjust the position of the widget relative to the top of the screen. This helps in customization of the user's home screen. When no widget is loaded, you can still use this slider to set the position from which your apps start populating.

**Comment by niakoi1** [comment]
#response_container_BBPPID{font-family: initial; font-size:initial; color: initial;} I can't find this slider when no widget is loaded. It only exist when the  'resize' option of the widget is activated. Before implementing widget this option was not existing.BTW, these are minor issues. I love this launcher.                                                                                                                                             Sent via BlackBerry Hub+ Inbox for Android                                                                                                                                                                           From: notifications@github.comSent: August 16, 2019 11:20 PMTo: Niagara-Issues@noreply.github.comReply-to: reply@reply.github.comCc: vassiletropolski@gmail.com; mention@noreply.github.comSubject: Re: [8bitPit/Niagara-Issues] Widget support (#117)  @niakoi1 the clock going off the screen is not intended and will be fixed. The upper slider is used to adjust the position of the widget relative to the top of the screen. This helps in customization of the user's home screen. When no widget is loaded, you can still use this slider to set the position from which your apps start populating.

**Comment by Giancarlomerida** [comment]
Hi chat, is there a way to put more than 1 widget? 

If not... will be an option in the near future?

Just found this launcher, in my huawei mate 10 pro with the live earth 3d wallpaper, looks amazing. 

Love it so much, but i miss my typical 4 widgets :)

Have a great day people!

**Comment by niakoi1** [comment]
Ok, another widget issue:
If you do not have the clock enabled and add an widget and move the widget all the way up. Then enable the clock. It will stay off the screen and will not be visible. The same is valid even if you removed the widget before enabling the clock

**Comment by MarcAlx** [comment]
Don't know if it's the right place to submit that. It would be great to let user chose the position of the Widget (top of favorites, bottom of favorites) as some widgets are more "actions" than displaying information. 

For exemple the wikipedia (official) widget that open the app to perform a search is the kind of widget I want on my screen but not on top of it, I prefer it on the bottom easily accessible one hand.

PS I can create a feature suggestions if needed

**Comment by Methotrex8** [comment]
I think this is where I should post this suggestion since it's about widgets. Instead of a separate page or a dedicated section of the screen, can you just integrate widgets when you swipe-right on an app? So basically, the pop-up with notifications can accommodate a widget area. I'm thinking along the lines of having it as an alternative to notifications (because notifications and widgets--to me--seem redundant). 

Also, I'm looking at my widgets now, and the only thing that I want is a quick way to see a calendar, so another suggestion is to put a swipe-right gesture on the date to launch a calendar widget of the person's choosing. I think another person above me also wants a weather widget, so that can also be implemented on the current weather icon. Am I making sense? Hahaha!

**Comment by victorxls** [comment]
Why dont you create a top horizontal scroll for widgets/extra homecreen pages. All my widgets could be accomodated without interfering with rest of the interface
![PicsArt_09-15-11 20 50](https://user-images.githubusercontent.com/55340872/64917148-41831580-d7ab-11e9-9ddd-aabedc70c6e1.png)

**Comment by 8bitPit** [comment]
@victorxls unfortunately this wouldn't work well with widgets of different size.

**Comment by niakoi1** [comment]
@8bitPit the widgets bug still exists.
To replicate:
1. Remove clock
2. Add a widget
3. Push it all the way up
4. Remove the widget
5. Add clock
Expected result: clock appears at the top
Actual result: clock stays off the screen

It would be nice to have it fixed :)

**Comment by TheIntellectualBadAss401** [comment]
I get to this is like a one-handed type of deal with the home screen and everything but I do feel that it needs a sense of urgency when it comes to widgets so I was thinking it would probably be a good idea to build a place that we just wherever you want you on the home screen but just as long as in fact it doesn't clash with the actual one handed operation in that sense when I'm basically indicating is on the home screen itself would be a good idea to have widgets allowing different areas but when you actually go to the side to go through and try to find the apps with the lettering the widgets disappear until you go back to the home screen this would be like an ideal type of theme for the most partand I don't believe it would actually mess up anything as far as one-handed operation goes at that point also just wanted to get my feedback. Thanks

**Comment by TXDLLL** [comment]
The drop-down hover  , open a new windows,like wechat

**Comment by thor9898** [comment]
> I was thinking a page to the right with a list of active widgets would be cool.

I'm really really into this if this ever happens I'm not changing launcher because I need more than one widget because of my profession.
I truly appreciate your work and it deserves the money

**Comment by 8bitPit** [comment]
Thank you very much @thor9898. Adding a widget page is not planned, but there may be other ways to add multiple widgets. For more information please read the following comment or visit https://github.com/8bitPit/Niagara-Issues/issues/1165.
> I am not really a fan of widget pages. The only thing widgets have in common there is, well, they’re all widgets. So why should they be grouped together? Showing a weather widget below the clock and accessing your calendar agenda via a swipe / by tapping on a button makes more sense to me. Also, widgets often tend to use much ram. The concept of a widget list would kind of endorse user to cluster the page full of widgets. A swipe menu shows only a single widget at once, so not every widget has to be loaded into ram when you want to access one.
That’s why I’m planning to add the following integration:
You can place a single widget below the clock or replace it for showing weather and / or time, or an at a glance widget.
More widgets can be accessed by swiping right on an app.
Another idea would be that you can customize the google search button at the bottom right, so it can also launch a widget, e.g. your calendar. [[Source](https://github.com/8bitPit/Niagara-Issues/issues/117#issuecomment-464157151)]

**Comment by tundratier** [comment]
Just some feedback that I will uninstall Niagara in favor of my old launcher because of this "missing" feature.

> The only thing widgets have in common there is, well, they’re all widgets. So why should they be grouped together? 

Because that is **the point** of widgets. I can have a detailed view of today's weather, my ToDo's, and the state of my Smart Home all by just waking up my phone. At a glance. It's the reason why most complex projects have dashboard views to give you an overview across all systems. Same principle applies to app widgets.
Using stacks or popup widgets is too much of a hassle and prevents me from getting a quick info without moving any finger, at all.

If you don't want this functionality because it doesn't fit in your concept, that's fine, of course. Your product doesn't have to fit all needs. Just wanted to let you know that you're loosing customers because of something that should be easily customizable by the user.

**Comment by niakoi1** [comment]
Not sure why I'm getting this email chain.
